chainstate/tx-verifier/src/transaction_verifier/input_output_policy/mod.rs

```diff
@@ -0,0 +1,356 @@
+            TxOutput::DelegateStaking(..) => true,
+        })
+        .count();
+
+    let spend_share_outputs_count = outputs
+        .iter()
+        .filter(|output| match output {
+            TxOutput::Transfer(..)
+            | TxOutput::Burn(..)
+            | TxOutput::CreateStakePool(..)
+            | TxOutput::ProduceBlockFromStake(..)
+            | TxOutput::CreateDelegationId(..)
+            | TxOutput::DelegateStaking(..) => false,
+            TxOutput::LockThenTransfer(..) => true,
+        })
+        .count();
+
+    delegation_input
+        && delegation_outputs_count < 2
+        && spend_share_outputs_count == outputs.len() - delegation_outputs_count
+}
+
+fn is_valid_any_to_any_combination_for_tx(
+    inputs_utxos: &[TxOutput],
+    outputs: &[TxOutput],
+) -> Result<(), ConnectTransactionError> {
+    let valid_inputs = are_inputs_valid_for_tx(inputs_utxos);
+    ensure!(valid_inputs, ConnectTransactionError::InvalidInputTypeInTx);
+    let valid_outputs = are_outputs_valid_for_tx(outputs);
+    ensure!(
+        valid_outputs,
+        ConnectTransactionError::InvalidOutputTypeInTx
+    );
+    Ok(())
+}
+
+fn are_inputs_valid_for_tx(inputs_utxos: &[TxOutput]) -> bool {
+    inputs_utxos.iter().all(|input_utxo| match input_utxo {
+        TxOutput::Transfer(..) | TxOutput::LockThenTransfer(..) => true,
+        TxOutput::Burn(..)
+        | TxOutput::CreateStakePool(..)
+        | TxOutput::ProduceBlockFromStake(..)
+        | TxOutput::CreateDelegationId(..)
+        | TxOutput::DelegateStaking(..) => false,
+    })
+}
+
+fn are_outputs_valid_for_tx(outputs: &[TxOutput]) -> bool {
+    let valid_outputs_types = outputs.iter().all(|output| match output {
+        TxOutput::Transfer(..)
+        | TxOutput::LockThenTransfer(..)
+        | TxOutput::Burn(..)
+        | TxOutput::CreateStakePool(..)
+        | TxOutput::CreateDelegationId(..)
+        | TxOutput::DelegateStaking(..) => true,
+        TxOutput::ProduceBlockFromStake(..) => false,
+    });
+
+    let is_stake_pool_unique = outputs
+        .iter()
+        .filter(|output| match output {
+            TxOutput::Transfer(..)
+            | TxOutput::LockThenTransfer(..)
+            | TxOutput::Burn(..)
+            | TxOutput::ProduceBlockFromStake(..)
+            | TxOutput::CreateDelegationId(..)
+            | TxOutput::DelegateStaking(..) => false,
+            TxOutput::CreateStakePool(..) => true,
+        })
+        .count()
+        < 2;
+
+    let is_create_delegation_unique = outputs
+        .iter()
+        .filter(|output| match output {
+            TxOutput::Transfer(..)
+            | TxOutput::LockThenTransfer(..)
+            | TxOutput::Burn(..)
+            | TxOutput::CreateStakePool(..)
+            | TxOutput::ProduceBlockFromStake(..)
+            | TxOutput::DelegateStaking(..) => false,
+            TxOutput::CreateDelegationId(..) => true,
+        })
+        .count()
+        < 2;
```

Comment:
Instead of `.count() < 2`, you can use `.at_most_one().is_ok()`. That has the advantage of exiting early if the second matching item is hit. Also avoids the slightly weird formatting on the comparison.

Reply:
Also there could be some sort of early exit if one of the conditions fail. One hack to do it would be to turn the conditions into lambdas and change the last line to:

```rust
    valid_outputs_types() && is_stake_pool_unique() && is_create_delegation_unique()
```

Not sure about the readability of that though

Reply:
Similar points apply in `is_valid_delegation_spending`